ci/check_typeignore.py

```diff
@@ -0,0 +1,67 @@
+#!/usr/bin/env python
+
+"""This script checks that no files in our source code have a "#type: ignore" comment without explicitly indicating the
+reason for the ignore. This is to ensure that we don't accidentally ignore errors that we should be fixing."""
+
+
+import os
+import re
+import pathlib
+import sys
+from typing import List
+
+import click
+
+# Regex to match a "# type: ignore" comment not followed by a reason.
+TYPE_IGNORE_COMMENT = re.compile(r"#\s*type:\s*ignore\s*(?![^\[]*\[)")
+
+# Regex to match a "# type: ignore[<reason>]" comment.
+TYPE_IGNORE_REASON_COMMENT = re.compile(r"#\s*type:\s*ignore\[(?P<reason>.*)\]")
+
+
+def check_file(file: pathlib.Path):
+    """Checks that the given file has no "# type: ignore" comments without a reason."""
+    with open(file, "r") as f:
+        for i, line in enumerate(f):
+            if TYPE_IGNORE_COMMENT.search(line):
+                raise ValueError(f"{file}:{i+1}: Found a '# type: ignore' comment without a reason.")
+
+            if search := TYPE_IGNORE_REASON_COMMENT.search(line):
+                reason = search.group("reason")
+                if reason == "":
+                    raise ValueError(f"{file}:{i+1}: Found a '# type: ignore[]' comment without a reason.")
+
+
+def check_files(files: List[pathlib.Path]):
+    """Checks that the given files have no "# type: ignore" comments without a reason."""
+    for file in files:
+        check_file(file)
+
+
+def get_files_to_check(root_dir: pathlib.Path) -> List[pathlib.Path]:
+    """Returns a list of files that should be checked for "# type: ignore" comments."""
+    # Get the list of files that should be checked.
+    files = []
+    for root, _, filenames in os.walk(root_dir):
+        for filename in filenames:
+            if filename.endswith(".py"):
+                files.append(pathlib.Path(root) / filename)
+
+    return files
+
+
+@click.command()
+@click.option("--root-dir", type=click.Path(exists=True, file_okay=False, dir_okay=True), default="src")
+def main(root_dir: str):
+    """Checks that no files in our source code have a "#type: ignore" comment without explicitly indicating the
+    reason for the ignore. This is to ensure that we don't accidentally ignore errors that we should be fixing."""
+    files = get_files_to_check(pathlib.Path(root_dir))
+    try:
+        check_files(files)
+    except ValueError as e:
+        print(e)
```

Comment:
This conversation was resolved but still just doing `print(e)`?

If there's a reason you want to keep it as-is then that's OK but good to leave a comment explaining why.

In general good to avoid resolving comments unless the underlying issue is definitely addressed. Otherwise it makes code reviews a lot harder (I can't just look at the diff since last review and outstanding comments, actually have to track what the issues from the last review were).

Reply:
Apologies about this, I re-wrote this a few times and in my mind I thought it had been resolved. Looking at it now I see I didn't leave a comment about it. 

My opinion was that since all the error messages that can arise are the same and come from the same line, printing the trace back doesn't really add much information, only if it's a bug in the script. 

Do you agree?

Reply:
Ah, you're imagining this is just the ValueError raised in `check_file`?

I think that's a bit of an assumption -- `ValueError` is used all over Python standard library, it could be raised for a reason you didn't expect.

If you made a custom exception (maybe inheriting from `ValueError` similar to in the CI notebook cleaning) and just caught and printed that I'd be OK with it.

Reply:
I thought I was using a custom exception? Must have accidentally overwritten the changes. I'll add that in and mark it as resolved if you're happy with that.

Reply:
Yep, changing to custom exception and marking this as resolved is OK by me.